openadapt/capture/_macos.py

```diff
@@ -1,23 +1,66 @@
+# flake8: noqa # type: ignore # noqa
 """Allows for capturing the screen and audio on macOS.
 
 This is based on: https://gist.github.com/timsutton/0c6439eb6eb1621a5964
 
 usage: see bottom of file
 """
+
+from ctypes import POINTER, c_char_p, c_int, c_void_p, cdll
 from datetime import datetime
 from sys import platform
 import os
 
-from Foundation import NSURL, NSObject  # type: ignore # noqa
-from Quartz import CGMainDisplayID  # type: ignore # noqa
-import AVFoundation as AVF  # type: ignore # noqa
-import objc  # type: ignore # noqa
+from Foundation import NSObject, NSLog, NSAutoreleasePool
+from Quartz import CIImage
+from CoreMedia import CMSampleBufferGetImageBuffer
+from Quartz.CoreGraphics import CGMainDisplayID
+from AppKit import NSBitmapImageRep, NSImage, NSPNGFileType, NSImageCompressionFactor
+import AVFoundation as AVF
+import objc
+
+
+NULL_PTR = POINTER(c_int)()
+
+_lib = cdll.LoadLibrary("/usr/lib/system/libdispatch.dylib")
+_dispatch_queue_create = _lib.dispatch_queue_create
+_dispatch_queue_create.argtypes = [c_char_p, c_void_p]
+_dispatch_queue_create.restype = c_void_p
+
+
+def dispatch_queue_create(name):
+    """Creates a new dispatch queue to which you can submit blocks."""
+    # https://developer.apple.com/documentation/dispatch/1453030-dispatch_queue_create
+    b_name = name.encode("utf-8")
+    c_name = c_char_p(b_name)
+    queue = _dispatch_queue_create(c_name, NULL_PTR)
 
-from openadapt import config
+    return queue
+
+
+class VideoDelegate(NSObject):
+    """Delegate for the video capture."""
+
+    def captureOutput_didOutputSampleBuffer_fromConnection_(
+        self, output, sampleBuffer, connection
+    ):
+        """Capture the output of the video."""
+
+        # NSLog("Delegate method called")
+        imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer)
+        ciImage = CIImage.imageWithCVImageBuffer_options_(imageBuffer, None)
+        bitmapRep = NSBitmapImageRep.alloc().initWithCIImage_(ciImage)
+        dict = {NSImageCompressionFactor: 1.0}
+        pngData = bitmapRep.representationUsingType_properties_(NSPNGFileType, dict)
+        if not os.path.exists("captures"):
+            os.makedirs("captures")
+        pngData.writeToFile_atomically_(
+            os.path.join("captures", "frame{}.png".format(datetime.now())), False
+        )
```

Comment:
What do you think about wrapping this in a condition, e.g. `if save_images:`, with `save_images=False` as a default argument to this function?

Reply:
Yeah that seems fine, just wanted to have something to show it works

Reply:
[3a4deff](https://github.com/OpenAdaptAI/OpenAdapt/pull/585/commits/3a4deff0d06d52c68466a2f489d12d3f326c5051)